[DOC] Guidelines for downgrade

PaulRMellor opened the pull request:
Signed-off-by: prmellor <pmellor@redhat.com>

**Documentation**

New section in the _Deploying and Upgrading_ guide to provide guidance on how to revert to previous versions of Operators and Kafka. 

Described in two stages like the upgrade: Cluster Operator, then Kafka downgrade (optionally). The CO procedure reverses the upgrade steps. Downgrade of Kafka mentions reverting `genericKafkaListener` as a step (but no mention of API versions).

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

PaulRMellor commented:
When ready to merge, please cherry-pick to 0.20

scholzj commented on documentation/assemblies/upgrading/assembly-downgrade-kafka-versions.adoc:
```diff
@@ -1,20 +1,13 @@
-* Log message format
-* ZooKeeper
-
+//Version constraints on the downgrade
```
Should this still be here?

PaulRMellor replied:
Note on the file included in the assembly

laidan6000 commented on documentation/assemblies/upgrading/assembly-downgrade-kafka-versions.adoc:
```diff
@@ -1,20 +1,13 @@
+[id='assembly-downgrade-kafka-versions-{context}']
 = Downgrading Kafka
 
 Kafka version downgrades are performed using the Cluster Operator.
```
_Suggested change:_

performed by the Cluster Operator.

laidan6000 commented on documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:
```diff
@@ -8,30 +8,11 @@
 
 For the `Kafka` resource to be downgraded, check:
 
+* [IMPORTANT] xref:com-target-downgrade-version-{context}[Compatibility of Kafka versions].
```
I'm not sure about the square bracket convention here. Perhaps bold is better?

PaulRMellor replied:
I tried the bold, but thought just the capital was enough in the end. We use a similar format elsewhere, like with _ADVANCED OPTION:_ for callouts

![image](https://user-images.githubusercontent.com/47596553/99826905-e0926280-2b50-11eb-83d1-6a95f8ec5224.png)

laidan6000 commented on documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:
```diff
@@ -8,30 +8,11 @@
 
+* [IMPORTANT] xref:com-target-downgrade-version-{context}[Compatibility of Kafka versions].
 * The Cluster Operator, which supports both versions of Kafka, is up and running.
 * The `Kafka.spec.kafka.config` does not contain options that are not supported in the version of Kafka you are downgrading to.
```
_Suggested change:_

Use the same phrase in lines 17 and 18:

Kafka version being downgraded to

PaulRMellor replied:
:+1:  I changed both to: _by the Kafka version being downgraded to._

laidan6000 commented on documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:
```diff
@@ -52,4 +52,4 @@ You now have an updated Cluster Operator, but the version of Kafka running in th
 .What to do next
 
-Following the Cluster Operator upgrade, you can perform a xref:assembly-upgrading-kafka-versions-str[Kafka upgrade].
+Following the Cluster Operator upgrade, you can perform a xref:assembly-upgrading-kafka-versions-str[Kafka downgrade].
```
I would just remove what to do next from this module.

PaulRMellor replied:
Assuming you mean just the title. Removed.

laidan6000 replied:
I actually meant this whole thing! Why would a user downgrade Kafka after they've upgraded the Cluster Operator?

laidan6000 commented on documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:
```diff
@@ -0,0 +1,51 @@
+
+.What to do next
+
+Following the Cluster Operator downgrade, you can perform a xref:assembly-downgrade-kafka-versions-{context}[Kafka downgrade].
```
You can or you must?

PaulRMellor replied:
Can, as in it's now possible after the CO downgrade.

laidan6000 reviewed (changes requested):
New procedure looks good. Made a few comments.

tombentley commented on documentation/assemblies/upgrading/assembly-downgrade.adoc:
```diff
@@ -0,0 +1,26 @@
+. Revert your Cluster Operator to the previous Strimzi version.
+** xref:proc-downgrade-cluster-operator-{context}[]
+
+. Downgrade all Kafka brokers and client applications to the previous Kafka version.
```
This is not correct. The operator currently takes charge of the interbroker protocol, and updates it automatically. That means that the user cannot safely downgrade the brokers, even if they didn't complete the upgrade. I.e. once they've started to upgrade the Kafka cluster then they can't currently safely back out. They can upgrade and downgrade the operator.

PaulRMellor replied:
@tombentley Thanks Tom. I thought we could perform the downgrade if interbroker versions were the same. So you recommend we drop the section on downgrading Kafka entirely for now?

tombentley replied:
@PaulRMellor yeah, it would work like that if the operator didn't bump the interbroker version for you. IOW once you've changed `Kafka.spec.kafka.version` you've essentially changed the interbroker version, and Kafka downgrade is no longer possible. So, yes we should drop the section on downgrading Kafka for the time being.

PaulRMellor commented:
To reflect the current situation with the interbroker protocol and `Kafka.spec.kafka.version`, 
I've dropped assemblies/upgrading/assembly-downgrade.adoc from the /deploying master.adoc, which keeps the Kafka downgrade content out of the build. I've added the _Downgrading the Cluster Operator to a previous version_ procedure directly to the /deploying master.adoc. We can reintroduce the Kafka downgrade sections when appropriate. I've moved the warning on not being able to revert the Kafka upgrade in the _Upgrading Kafka brokers and client applications_ procedure to the step where change `Kafka.spec.kafka.version`.

PaulRMellor commented:
@tombentley @ppatierno @scholzj - looking for any more comments on **_proc-downgrade-cluster-operator.adoc_** and
**_proc-upgrade-brokers-newer-kafka.adoc_** (The other files are left out the doc until we can perform a kafka downgrade)

tombentley commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -4,15 +4,40 @@
+You cannot revert to the previous Kafka version if the `inter.broker.protocol.version` differs, or messages have been added
+to message logs that use a newer `log.message.format.version`.
+
+The `inter.broker.protocol.version` determines the schema used in the messages written to `__consumer_offsets`.
```
It's not _just_ this, it's used to gate other persistent data too (such as the versions of JSON persisted in zookeeper).

PaulRMellor replied:
How about:

> The `inter.broker.protocol.version` determines the schemas used for persistent metadata stored by the broker, such as the schema for messages written to `__consumer_offsets`.

Unless we want to go into more detail?

PaulRMellor replied:
This section is being left out for now

tombentley commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -4,15 +4,40 @@
+to message logs that use a newer `log.message.format.version`.
+
+The `inter.broker.protocol.version` determines the schema used in the messages written to `__consumer_offsets`.
+If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the broker will not understand the messages.
```
While that's true, it doesn't inform the reader about what the consequences of that will be. We should be clear that, in general, it  basically results in a "broker cluster". The rest of this concept then goes into detailed talk about the log message format version which could easily mislead people into thinking this sentence is not really a problem. I think we should probably omit the following sentences because they're mostly irrelevant due to a broker cluster generally being a "game over" situation.

PaulRMellor replied:
We're leaving this section out for now. I'm not sure how much we will want to retain when this situation is rectified?

PaulRMellor replied:
I've updated this sentence to:

> Therefore, if you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work.

scholzj commented on documentation/assemblies/upgrading/assembly-downgrade.adoc:
```diff
@@ -0,0 +1,26 @@
+[id='assembly-downgrade-{context}']
+= Downgrading Strimzi
+
+If you are encountering issues with the latest version of Strimzi,
```
The `latest` version of Strimzi sounds weird given the documentation is for the `latest` Strimzi version for a month or two but then it is for some older version but might be still use. Can we maybe rephrase it to say something like _If you are encountering issues with the version of Strimzi you upgraded to_?

scholzj commented on documentation/assemblies/upgrading/assembly-downgrade.adoc:
```diff
@@ -0,0 +1,26 @@
+If you are encountering issues with the latest version of Strimzi,
+you can revert your installation to the previous version.
+
+As with the Strimzi upgrade process, you can perform your downgrade in two stages:
+
+. Revert your Cluster Operator to the previous Strimzi version.
+** xref:proc-downgrade-cluster-operator-{context}[]
+
+. Downgrade all Kafka brokers and client applications to the previous Kafka version.
+** xref:assembly-downgrade-kafka-versions-{context}[]
```
I think this will be confusing ... saying it happens in two stages for me gives it a particular order. But the order is not so simple. For example, I'm running Strimzi 0.19.0 with Kafka 2.5. And I upgrade first to Strimzi 0.20 + Kafka 2.5 and then to Kafka 2.6. In that case, the order here is actually first to downgrade Kafka 2.6 to 2.5 if still possible and then to downgrade Strimzi 0.20 to 0.19.

PaulRMellor replied:
I've simplified to:

```
You can perform a downgrade to:

. Revert your Cluster Operator to the previous Strimzi version.
** xref:proc-downgrade-cluster-operator-{context}[]

. Downgrade all Kafka brokers and client applications to the previous Kafka version.
** xref:assembly-downgrade-kafka-versions-{context}[]
```

PaulRMellor replied:
This section won't be in the doc yet

scholzj commented on documentation/modules/upgrading/proc-downgrade-brokers-older-kafka.adoc:
```diff
@@ -70,6 +51,11 @@ NOTE: You must format the value of `log.message.format.version` as a string to p
 +
 See xref:con-versions-and-images-str[]
 
+.. If you changed the Kafka listener configuration to use the array format of the `GenericKafkaListener` schema,
+revert the `Kafka.spec.kafka.listeners` to use the old format.
++
+See xref:con-upgrade-listeners-str[]
```
This has to be part of the Cluster Operator downgrade. It does not have to be changed with Kafka downgrade. The operator supports the new format with any Kafka version it supports.

I also wonder if we should make it more generic: _Revert any changes to the CRs_ or something. As there are more new things in each release. E.g. rack awareness in Connect, host aliases and other things are not supported after the operator downgrade and need to be removed. So making it more generic would be IMHO a plus.

PaulRMellor replied:
Removed from this procedure and added a new generic procedure to the CO downgrade procedure.

![image](https://user-images.githubusercontent.com/47596553/100908609-98196400-34c3-11eb-8d32-0ff2c022d596.png)

scholzj commented on documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:
```diff
@@ -0,0 +1,50 @@
+[id='proc-downgrade-cluster-operator-{context}']
+= Downgrading the Cluster Operator to a previous version
+
+If you are encountering issues with the latest version of Strimzi,
```
Same as earlier ... the _latest version_ sounds weird and confusing here.

PaulRMellor replied:
Changed to:

```
If you are encountering issues with Strimzi,
you can revert your installation.
```

scholzj commented on documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:
```diff
@@ -67,6 +68,10 @@ spec:
 <2> This remains at the current version
+--
++
+WARNING: You cannot downgrade Kafka once `spec.kafka.version` has changed.
```
Is this because of the `inter.broker.protocol.version`? Or why is it the case?

tombentley replied:
`inter.broker.protocol.version` indeed

PaulRMellor replied:
Changed to:

> WARNING: You cannot downgrade Kafka once `spec.kafka.version` has changed, as the `inter.broker.protocol.version` for the new Kafka version will differ.  The protocol version determines the schemas used for persistent metadata stored by the broker, including messages written to `__consumer_offsets`. The cluster will not understand the messages.

scholzj replied:
Are we cherry-picking this for 0.20 release branch? If not, maybe we should think about this a bit more to improve the experience of this before we do 0.21 and before we merge this.

PaulRMellor replied:
Yes. Once it's cherry-picked, we can bring the kafka downgrade sections back in and improve the chapter a whole.

laidan6000 commented on documentation/deploying/master.adoc:
```diff
@@ -20,3 +20,5 @@ include::assemblies/deploying/assembly-deploy-verify.adoc[leveloffset=+1]
 //Upgrading the deployment
 include::assemblies/upgrading/assembly-upgrade.adoc[leveloffset=+1]
+//Downgrading the deployment
+include::modules/upgrading/proc-downgrade-cluster-operator.adoc[leveloffset=+1]
```
Just checking that the correct module is included in the `master.adoc`?

PaulRMellor replied:
For the time being, yes. There's a new module for downgrading, but we're concentrating only on the CO downgrade for the time being. Other files exists, but won't be included yet.

laidan6000 commented on documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:
```diff
@@ -0,0 +1,52 @@
+include::snip-cluster-operator-namespace-sed.adoc[]
++
+.. If you modified one or more environment variables in your existing Cluster Operator `Deployment`, edit the
+`install/cluster-operator/060-Deployment-cluster-operator.yaml` file to use those environment variables.
```
In Strimzi 0.19.0, this file is named `050-Deployment-strimzi-cluster-operator.yaml`.

PaulRMellor replied:
Still 060, but new name.

laidan6000 commented on documentation/modules/upgrading/proc-downgrade-cluster-operator.adoc:
```diff
@@ -0,0 +1,52 @@
++
+The image tag shows the new Strimzi version followed by the Kafka version. For example, `_NEW-STRIMZI-VERSION_-kafka-_CURRENT-KAFKA-VERSION_`.
+
+You now have an updated Cluster Operator.
```
_Suggested change:_

Your Cluster Operator was downgraded to the previous version.

laidan6000 commented on documentation/modules/upgrading/proc-upgrade-brokers-newer-kafka.adoc:
```diff
@@ -67,6 +68,9 @@ spec:
 <2> This remains at the current version
+--
++
+WARNING: You cannot downgrade Kafka once `spec.kafka.version` has changed, as the `inter.broker.protocol.version` for the new Kafka version will differ. The protocol version determines the schemas used for persistent metadata stored by the broker, including messages written to `__consumer_offsets`. The cluster will not understand the messages.
```
```suggestion
WARNING: You cannot downgrade Kafka after `spec.kafka.version` has changed, because the `inter.broker.protocol.version` for the new Kafka version will differ. The inter-broker protocol version determines the schemas used for persistent metadata stored by the broker, including messages written to `__consumer_offsets`. The downgraded cluster will not understand the messages.
```

laidan6000 commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -4,15 +4,40 @@
 
 [id='com-target-downgrade-version-{context}']
```
Typo in module ID: con.

laidan6000 commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -4,15 +4,40 @@
+to message logs that use a newer `log.message.format.version`.
+
+The `inter.broker.protocol.version` determines the schemas used for persistent metadata stored by the broker, such as the schema for messages written to `__consumer_offsets`.
+If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work as it will not understand the messages.
```
```suggestion
If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work because it will not understand the messages.
```

laidan6000 reviewed (changes requested):
Much clearer overall. A few questions sprung to mind on a second reading.

laidan6000 reviewed (approved):
Changes look good.

PaulRMellor commented:
Kafka downgrade sections reintroduced, following the work to make specifying `inter.broker.protocol.version` explicit in https://github.com/strimzi/strimzi-kafka-operator/pull/4044 and https://github.com/strimzi/strimzi-kafka-operator/pull/4058

PaulRMellor commented on documentation/modules/upgrading/proc-upgrade-cluster-operator.adoc:
```diff
@@ -42,14 +44,12 @@ Wait for the rolling updates to complete.
 kubectl get po my-cluster-kafka-0 -o jsonpath='{.spec.containers[0].image}'
```
@tombentley - we do the check on the Kafka pod for the CO upgrade?

PaulRMellor replied:
It was something that @michalxo mentioned when reviewing the procedure. He wondered why we don't do the verification on the CO pod. 

```
kubectl get pod <cluster-operator-pod-name> -o jsonpath='{.spec.containers[0].image}'
        quay.io/strimzi/operator:latest%
```

tombentley replied:
Still not understanding this. `get po` and `get pod` are the same thing. Or are you referring to the `quay.io/strimzi/operator:latest%` bit?

michalxo replied:
Looking at my comment in Jira, I think the latter is what I meant to suggest for a change ->
``` wrong pod used in command?. As we upgraded strimzi CO and not kafka (that did not change)

$ kubectl get pod <cluster-operator-pod-name> -o jsonpath='{.spec.containers[0].image}'
        quay.io/strimzi/operator:latest%
```

scholzj replied:
IMHO the upgrade of the operator is successful after all operand pods roll to the new images. So checking the image in the Kafka pods makes more sense than checking it in the Cluster Operator pod.

tombentley replied:
Ah, finally I understand what's being asked here! (Sorry if I was being dense). Yes, checking on the operand pods makes more sense.

scholzj replied:
@PaulRMellor If you do any edits to this, I think using the full name - `kubectl get pod` makes more sense in documentation.

PaulRMellor replied:
@tombentley @scholzj @michalxo  -- Sorry for the confusion and thanks for the comments. I've left the commands as is, but changed to `pod` rather than `po`

PaulRMellor commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -2,17 +2,42 @@
+
+[source,yaml,subs=attributes+]
+----
+apiVersion: v1alpha1
```
@scholzj api version check on the upgrade and downgrade procedures. We have `v1alpha1` in our examples

scholzj replied:
Aren't we normally using some variable? It should definitely be `v1beta1`.

PaulRMellor replied:
I've replaced with the {KafkaApiVersion} variable.

scholzj commented:
@tombentley Does this look good now? Were your comments addressed?

tombentley commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -2,17 +2,42 @@
+to message logs that use a newer `log.message.format.version`.
+
+The `inter.broker.protocol.version` determines the schemas used for persistent metadata stored by the broker, such as the schema for messages written to `__consumer_offsets`.
+If you downgrade to a version of Kafka that uses an earlier `inter.broker.protocol.version`, the cluster will not work because it will not understand the messages.
```
An "earlier IBP" isn't quite the right language. Each version of Kafka supports a range of IBPs, so "earlier" is a little ambiguous (it could be taken to mean earlier in the range of IBPs supported by a particular version). 

```suggestion
If you downgrade to a version of Kafka that doesn't understand an `inter.broker.protocol.version` that has (ever) been previously used in the cluster the broker will encounter data they cannot understand.
```

tombentley commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -2,17 +2,42 @@
-* If the target downgrade version of Kafka has the same `log.message.format.version` as the current version, the Cluster Operator downgrades by performing a single rolling restart of the brokers.
-* If the target downgrade version of Kafka has a different `log.message.format.version`, downgrading is only possible if the running cluster has _always_ had `log.message.format.version` set to the version used by the downgraded version.
-+
+You cannot revert to the previous Kafka version if the `inter.broker.protocol.version` differs, or messages have been added
```
This sentence on its own (I know is clarified in a couple of sentences time) might make the reader think they can just change the IBP first. 

```suggestion
You cannot revert to the previous Kafka version if that version don't support any of the `inter.broker.protocol.version` settings which have ever been used in that cluster, or messages have been added
```

PaulRMellor commented:
@tombentley - do you have any more comments?

tombentley commented on documentation/modules/upgrading/con-downgrade-target-version.adoc:
```diff
@@ -2,17 +2,42 @@
-* If the target downgrade version of Kafka has the same `log.message.format.version` as the current version, the Cluster Operator downgrades by performing a single rolling restart of the brokers.
-* If the target downgrade version of Kafka has a different `log.message.format.version`, downgrading is only possible if the running cluster has _always_ had `log.message.format.version` set to the version used by the downgraded version.
-+
+You cannot revert to the previous Kafka version if that version don't support any of the `inter.broker.protocol.version` settings which have ever been used in that cluster,
```
```suggestion
You cannot revert to the previous Kafka version if that version don't support any of the `inter.broker.protocol.version` settings which have _ever been used_ in that cluster,
```

scholzj commented:
@PaulRMellor Looks like the CI is failing:

```
./.azure/scripts/check_docs.sh
Avoid 'nt contraction:
documentation/modules/upgrading/con-downgrade-target-version.adoc:16:If you downgrade to a version of Kafka that doesn't understand an `inter.broker.protocol.version` that has (ever) been previously used in the cluster the broker will encounter data they cannot understand.
ERROR: 1 docs problems found.
Makefile:138: recipe for target 'docu_check' failed
```
